File: merge-sort/merge_sort.jl
# coding: utf-8

## Implementation of the *quick-sort* algorithm

### Load the data:


numbers = readdlm("IntegerArray_example.txt",'\n',Int)' # notice transposition
numbers = {[n] for n=numbers}


### The functions:

function merge_sort_binary(vec1,vec2)
    i      = 1
    j      = 1
    len1   = length(vec1)
    len2   = length(vec2)
    output = Int64[]
    inversions = 0
    
    
    while i+j <= len1+len2+1
        if i<len1+1 && j<len2+1     # this avoids BoundsError()
            if vec1[i] <= vec2[j]
                output = vcat(output,vec1[i])
                i+=1
            else
                output      = vcat(output,vec2[j])
                inversions += len1-i+1
                j+=1
            end
        else                        # this takes care of boundary cases
            if j==len2+1
                output = vcat(output,vec1[i])
                i+=1
            else
                output      = vcat(output,vec2[j])
                inversions += len1-i+1
                j+=1
            end
        end
    end
    
    return inversions, output
end






function merge_sort_iterative(cellarray)
    len = length(cellarray)
    odd = len % 2
    inversions = 0
    
    output = {}
    
    for i=1:2:len-1
        newinversions, newarray  = merge_sort_binary( cellarray[i], cellarray[i+1] )
        output = vcat( output, {newarray} )
        inversions += newinversions
    end
    
    if odd == 1
        output = vcat( output, {cellarray[len]} )
    end
    
    return inversions, output
end






function merge_sort(cellarray)
    inversions = 0
    
    while length(cellarray) > 1
        newinversions, cellarray = merge_sort_iterative(cellarray)
        inversions += newinversions
    end
    
    return inversions, cellarray[1]
end



### The function call:

merge_sort(numbers)

File: merge-sort/merge_sort.cc
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
 
using namespace std;





//-----{ global variables }---------------------------------------------------//

string input_filename  = "IntegerArray_example.txt"; 
long   inversion_count = 0;

//----------------------------------------------------------------------------//






//-----{ function templates }-------------------------------------------------//

vector<vector<int>> get_numbers         ( string                filename    );

vector<int>         merge_sort          ( vector<vector<int>>   input       );

void                print_vector        ( vector<int>           input       );
void                print_vector_nested ( vector<int>           input       );

//----------------------------------------------------------------------------//







//---{ main }-----------------------------------------------------------------//

int main ()
{
    
    // load the numbers:
    vector<vector<int>> numbers = get_numbers( input_filename );
    
    // sort numbers:
    merge_sort(numbers);

    // print the number of inversions:
    cout << inversion_count << endl;

}

//----------------------------------------------------------------------------//











//==={ FUNCTION DEFINITIONS }=================================================//





//---{ load the data as a vector of integers }--------------------------------//

vector<vector<int>> get_numbers (string filename)
{
    ifstream file ( filename );
    vector<vector<int>> numbers;
    
    // check whether the file is open:
    if ( !file.is_open() ) 
    {
        cout << "Can't find the file \"" << filename << "\"." << endl; 
    }
    else 
    {
        int number;
        while (file >> number)              // read line from file
        {
            numbers.push_back({number});    // append to vector
        }
        
        file.close();
    }
    
    return numbers;
}

//----------------------------------------------------------------------------//










//---{ the binary merge-sort operation }--------------------------------------//

// This function takes two vectors and performs the binary merge-sort operation,
// so it returns a single sorted vector. This function forms the core of the
// algorithm.

vector<int> merge_sort_binary (vector<int> vec1, vector<int> vec2)
{
    vector<int> output;
    int i=0, j=0;                           // indices of vec1 and vec2, resp.
    int len1=vec1.size(), len2=vec2.size(); // length of vec1 and vec2
    int length = len1 + len2;               // length of the output vector
    
    
    // The condition in the following "if" clause is the only tricky bit of
    // this function. It checks whether or not the next element in output
    // should come from vec1.

    do
    {
        if ( (vec1[i]<=vec2[j] && i<len1) || j==len2 )
        {
            output.push_back(vec1[i]);  // take element from vec1
            i++;                        // advance to next entry in vec1
        }
        else 
        {
            output.push_back(vec2[j]);  // take element from vec2
            j++;                        // advance to next entry in vec1
            inversion_count += len1-i;  // increase number of inversions
        }

    }
    while ( i+j < length );
    
    
    return output;
}

//----------------------------------------------------------------------------//










//---{ the iterative merge-sort function }------------------------------------//

// This function iteratively applies the binary merge-sort operator to all 
// consecutive pairs of vectors. (the input is a vector of nested vectors)

vector<vector<int>> merge_sort_iterative (vector<vector<int>> input)
{
    
    vector<vector<int>> output;
    
    
    int length = input.size();
    
    for ( int i=0; i<length/2; i++ )
    {
        output.push_back(
            merge_sort_binary( input[2*i], input[2*i+1] )   // Append merged
        );                                                  // vector to output.
    }
    
    if ( length % 2 )                       // Append the remaining vector
        output.push_back(input[length-1]);  // in case # elements is odd.

    
    
    return output;
}

//----------------------------------------------------------------------------//









//---{ the recursive merge-sort call }----------------------------------------//

vector<int> merge_sort ( vector<vector<int>> input )
{
    inversion_count = 0;
    
    vector<vector<int>> output = input;
    
    while ( output.size() > 1 )                 // Recursively apply the 
        output = merge_sort_iterative(output);  // iterative merge-sort func.
    
    return output[0];           // return the first and only element of output
}

//----------------------------------------------------------------------------//










//---{ these functions are just printing shortcuts (unimportant) }------------//

void print_vector_base ( vector<int> input )
{
    cout << "{";
    
    for (unsigned int i=0; i<input.size(); i++)
        
        if (i<input.size()-1)
            cout << input[i] << ',';
        else
            cout << input[i];
            
    cout << "}";
}



void print_vector ( vector<int> input )
{
    print_vector_base(input);
    cout << endl;
}



void print_vector_nested ( vector<vector<int>> input )
{
    cout << "{ ";
    
    for (unsigned int i=0; i<input.size(); i++)
    {
        print_vector_base(input[i]);
        
        if (i<input.size()-1)
            cout << ", ";
        else
            cout << "";
    }

    cout << " }" << endl;
}

//----------------------------------------------------------------------------//






